### PR TITLE
Allow setting knock room visibility

### PR DESCRIPTION
**Pull request by charlynguyen**
We @nordeck are currently implementing the knock rooms behind the feature flag `feature_ask_to_join` (introduced in https://github.com/matrix-org/matrix-react-sdk/pull/11182).

This pull request allows setting knock room visibility.

- Add checkbox to handle the knock room visibility when creating rooms (CreateRoomDialog).
- Add checkbox to handle the knock room visibility when editing rooms (JoinRuleSettings).
- Introduce "Ask to join" button to spotlight options to directly prompt the request to join (SpotlightDialog).

Epic: https://github.com/vector-im/element-web/issues/18655

#### ~Is blocked by~

- ~https://github.com/matrix-org/matrix-js-sdk/pull/3673~

#### Relates to

- https://github.com/matrix-org/matrix-react-sdk/pull/11404
- https://github.com/matrix-org/matrix-react-sdk/pull/11353
  - https://github.com/matrix-org/matrix-js-sdk/pull/3647
- https://github.com/matrix-org/matrix-react-sdk/pull/11248
- https://github.com/matrix-org/matrix-react-sdk/pull/11182
- https://github.com/vector-im/element-web/pull/25923

#### CreateRoomDialog

CreateRoomDialog

#### JoinRuleSettings

JoinRuleSettings

#### SpotlightDialog

SpotlightDialog

## Checklist

-   [x] Tests written for new code (and old code if feasible)
-   [x] Linter and other CI checks pass
-   [x] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-react-sdk/blob/develop/CONTRIBUTING.md))

---
Here's what your changelog entry will look like:

## ✨ Features
 * Allow setting knock room visibility ([\#11464](https://github.com/matrix-org/matrix-react-sdk/pull/11464)). Contributed by @charlynguyen.

### CODE REVIEW

**Comment by charlynguyen** [comment]
FYI @daniellekirkwood 😄

File: src/components/views/dialogs/spotlight/SpotlightDialog.tsx [hunk 1]
```diff
@@ -538,10 +550,15 @@ const SpotlightDialog: React.FC<IProps> = ({ initialText = "", initialFilter = n
             metricsViaKeyboard: viaKeyboard,
             room_id: room.roomId,
             room_alias: room.roomAlias,
-            auto_join: room.autoJoin,
+            auto_join: room.autoJoin && !canAskToJoin(room.joinRule),
```
Comment (weeman1337) [inline src/components/views/dialogs/spotlight/SpotlightDialog.tsx:553]: What if I am allowed to join the room by the [`knock_restricted` rule](https://spec.matrix.org/v1.8/client-server-api/#mroomjoin_rules)?

**Review by weeman1337 — CHANGES_REQUESTED** [review]
Thank you for your contribution @charlynguyen . Please have a look at my comments:

We are using switches in most places. Can you replace the checkboxes with switches? Otherwise the UI would be (more) inconsistent:

| switch | checkbox |
| --- | --- |
| image | image |

File: src/components/views/dialogs/CreateRoomDialog.tsx [hunk 2]
```diff
@@ -215,6 +217,10 @@ export default class CreateRoomDialog extends React.Component<IProps, IState> {
         return result;
     };
 
+    private onIsPublicChange = (isPublic: boolean): void => {
+        this.setState({ isPublic });
```
Comment (richvdh) [inline src/components/views/dialogs/CreateRoomDialog.tsx:221]: for reference: I was confused about what this `isPublic` state was doing. The answer is: nothing. It appears to have been introduced by https://github.com/matrix-org/matrix-react-sdk/pull/5698, probably due to an incorrect merge resolution due to a conflict with https://github.com/matrix-org/matrix-react-sdk/pull/6212 (which removed it).

**Review by richvdh — CHANGES_REQUESTED** [review]
Some additional thoughts from me:

* This feels like three separate changes, which are only very tangentially related. I think it would be easier to review and merge if split into separate PRs.
*  We already have a switch to control whether a room is in the public room directory (on the "General" tab) - I think adding a second control on another tab will be confusing.

**Comment by weeman1337** [comment]
>     * We already have a switch to control whether a room is in the public room directory (on the "General" tab) - I think adding a second control on another tab will be confusing.

I've checked [the designs](https://www.figma.com/file/7TqjqdMBaPpm3IavKsMYu6/%5BBWI%5D-Knocking?type=design&node-id=367-57456&mode=design&t=UtlAGTyFZ6fFCBKt-0). This is how they suggested it. Doesn't mean we shouldn't open this up for discussion.

**Comment by richvdh** [comment]
> I've checked [the designs](https://www.figma.com/file/7TqjqdMBaPpm3IavKsMYu6/%5BBWI%5D-Knocking?type=design&node-id=367-57456&mode=design&t=UtlAGTyFZ6fFCBKt-0). This is how they suggested it. Doesn't mean we shouldn't open this up for discussion.

yeah... I'd like to understand the reasoning behind that decision. Duplicating the setting, but *only* for knockable rooms, seems confusing to me.

**Comment by daniellekirkwood** [comment]
I agree with this:
> We already have a switch to control whether a room is in the public room directory (on the "General" tab) - I think adding a second control on another tab will be confusing.

However, the switch being "point of need" at the "ask to join" part of settings makes sense to me. If we want to remove it from General, then maybe we explore that.

Having discussed the "Compoundifying" of Knocking with @nadonomy this morning we've agreed that it's not on our critical path and we're happy with the team at Nordeck to push forwards with the Knocking designs that we have provided them - then agreeing that when we get to Settings in the Compound project (Q4/Q1) we will review all the switches, toggles, pages and repeated info hollistically and remove them later.

**Comment by richvdh** [comment]
> Having discussed the "Compoundifying" of Knocking with @nadonomy this morning we've agreed that it's not on our critical path and we're happy with the team at Nordeck to push forwards with the Knocking designs that we have provided them

@daniellekirkwood can you just confirm my understanding on a few points:

 * You're happy for these controls to be checkboxes rather than switches
 * You're happy to duplicate the "include in public directory" switch to multiple room-settings tabs
 * You're happy for the setting to only appear for knockable rooms, and not (say) "public" rooms. (This seems *particularly* unintuitive to me)

**Comment by daniellekirkwood** [comment]
> You're happy for these controls to be checkboxes rather than switches

I'm happy for this PR to use the method that's been shared in the designs. Esp. knowing that it's likely we'll update it in the next 12 months as part of the compound project. Unless @nadonomy has a big blocker here.

> You're happy to duplicate the "include in public directory" switch to multiple room-settings tabs

Not happy but believe it's a temporary necessity as we don't have the answer to where it should like (nor the time/bandwidth to find the answer). The two switches should be linked though - if one is on, the other should reflect that state and vice versa. 

> You're happy for the setting to only appear for knockable rooms, and not (say) "public" rooms. (This seems particularly unintuitive to me)

Good point -- when someone selects Public it doesn't necessarily add it to the directory, right? Maybe this checkbox should appear there also... @nadonomy i would like support on that question...

**Comment by charlynguyen** [comment]
> Thank you for your contribution @charlynguyen . Please have a look at my comments:
> 
> We are using switches in most places. ~Can you replace the checkboxes with switches? Otherwise the UI would be (more) inconsistent.~ Update: [Looks like design suggests it this way](https://www.figma.com/file/7TqjqdMBaPpm3IavKsMYu6/%5BBWI%5D-Knocking?type=design&node-id=367-50166&mode=design&t=UtlAGTyFZ6fFCBKt-0).
> 
> switch	checkbox
> image	image

That was also my impression during the implementation, but I decided to stick to the screens.

File: src/components/views/dialogs/spotlight/SpotlightDialog.tsx (hunk 1 above)
Comment (charlynguyen) [reply src/components/views/dialogs/spotlight/SpotlightDialog.tsx:553]: Is `knock_restricted` already implemented in element-web? I consider it to be handled separately.

File: src/components/views/dialogs/CreateRoomDialog.tsx (hunk 2 above)
Comment (charlynguyen) [reply src/components/views/dialogs/CreateRoomDialog.tsx:221]: That was also my conclusion. So, I decided to give it a purpose.

**Comment by charlynguyen** [comment]
> It seems like this does match the intentions from the design team, but I'd still rather see it land as three separate pull requests rather than a single one. @charlynguyen would you mind breaking it up?

That is totally fine by me 🙂

File: src/components/views/dialogs/spotlight/SpotlightDialog.tsx (hunk 1 above)
Comment (weeman1337) [reply src/components/views/dialogs/spotlight/SpotlightDialog.tsx:553]: Okay

**Pull request closed without merging**